include/nbl/builtin/hlsl/subgroup/fft.hlsl

```diff
@@ -12,49 +12,94 @@ namespace hlsl
 namespace subgroup
 {
 
-// TODO: specialize subgorupShuffleXor to different Scalar types. As of right now this would only support 1-wide scalars (no vectors)
-template<typename Scalar, bool inverse>
-void FFT_loop(uint32_t stride, NBL_REF_ARG(complex_t<Scalar>) lo, NBL_REF_ARG(complex_t<Scalar>) hi) 
+// -----------------------------------------------------------------------------------------------------------------------------------------------------------------
+template<uint16_t K, bool Inverse, typename Scalar, class device_capabilities=void>
+struct FFT
 {
-    const vector <Scalar, 4> loHiPacked = {lo.real(), lo.imag(), hi.real(), hi.imag()};
-    vector <Scalar, 4> shuffledLoHiPacked = glsl::subgroupShuffleXor< vector <Scalar, 4> > (loHiPacked, stride);
-    lo.real(shuffledLoHiPacked.x);
-    lo.imag(shuffledLoHiPacked.y);
-    hi.real(shuffledLoHiPacked.z);
-    hi.imag(shuffledLoHiPacked.w);
-    // Get twiddle with k = subgroupID mod stride, N = 2 * stride
-    fft::DIX<Scalar, inverse>::radix2(fft::twiddle<Scalar, inverse>(glsl::gl_SubgroupID() & (stride - 1), stride << 1), lo, hi);   
-}
+    static void __call(NBL_REF_ARG(complex_t<Scalar>) lo, NBL_REF_ARG(complex_t<Scalar>) hi);
+};
+
+// ---------------------------------------- Radix 2 forward transform - DIF -------------------------------------------------------
 
-// When doing a 1D FFT of size N, we call N/2 threads to do the work, where each thread of index 0 <= k < N/2 is in charge of computing two elements
-// in the output array, those of index k and k + N/2. The first is "lo" and the latter is "hi"
-template<typename Scalar, uint32_t SubgroupSize, bool inverse>
-void FFT(NBL_REF_ARG(complex_t<Scalar>) lo, NBL_REF_ARG(complex_t<Scalar>) hi) 
+template<typename Scalar, class device_capabilities>
+struct FFT<2, false, Scalar, device_capabilities>
 {
-    const uint32_t doubleSubgroupSize = SubgroupSize << 1;
+    static void FFT_loop(uint32_t stride, NBL_REF_ARG(complex_t<Scalar>) lo, NBL_REF_ARG(complex_t<Scalar>) hi)
+    {
+        const bool topHalf = (glsl::gl_SubgroupInvocationID() & stride) != 0;
+        const vector <Scalar, 2> toTrade = topHalf ? vector <Scalar, 2>(lo.real(), lo.imag()) : vector <Scalar, 2>(hi.real(), hi.imag());
```

Comment:
btw `?` should work on regular structs, packing to/from `vector<Scalar,2>` to use in `subgroupShuffleXor` should happen after the ternary for readability

Reply:
does it not work on structs or what?

Reply:
ok I see that `?` not supported on structs